Creating clientAttributesCondition in some client policy breaks the login to the realm

### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

authentication

### Describe the bug

- Create some realm
- Go to `client policies` tab and create a condition of type `Client Attributes`
- Add some attributes to it (like "key1" with "value1" and "key2" with "value2") and save it. Note it is not needed to setup any profiles to the client policy
- Try to login to such a realm. The login page does not yet open as there is an error displayed to the user and there is an exception like this:

```
java.lang.IllegalArgumentException: Cannot construct instance of `java.util.LinkedHashMap` (although at least one Creator exists): no String-argument constructor/factory method to deserialize from String value ('[{"key":"aaa","value":"bbb"},{"key":"bbb","value":"ccc"}]')
 at [Source: UNKNOWN; byte offset: #UNKNOWN] (through reference chain: org.keycloak.services.clientpolicy.condition.ClientAttributesCondition$Configuration["attributes"])
	at com.fasterxml.jackson.databind.ObjectMapper._convert(ObjectMapper.java:4624)
	at com.fasterxml.jackson.databind.ObjectMapper.convertValue(ObjectMapper.java:4555)
	at org.keycloak.services.clientpolicy.ClientPoliciesUtil.getConditionProvider(ClientPoliciesUtil.java:454)
	at org.keycloak.services.clientpolicy.ClientPoliciesUtil.getEnabledClientPolicies(ClientPoliciesUtil.java:430)
	at org.keycloak.services.clientpolicy.DefaultClientPolicyManager.doPolicyOperation(DefaultClientPolicyManager.java:64)
	at org.keycloak.services.clientpolicy.DefaultClientPolicyManager.triggerOnEvent(DefaultClientPolicyManager.java:56)
	at org.keycloak.protocol.oidc.endpoints.AuthorizationEndpoint.process(AuthorizationEndpoint.java:136)
	at org.keycloak.protocol.oidc.endpoints.AuthorizationEndpoint.buildGet(AuthorizationEndpoint.java:115)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:154)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:118)
	at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:560)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:452)
	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:413)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:321)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:415)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:378)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:174)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:142)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:168)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:131)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:33)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:429)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:240)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:154)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:321)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:157)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:229)
	at org.jboss.resteasy.plugins.server.servlet.ServletContainerDispatcher.service(ServletContainerDispatcher.java:222)
	at org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher.service(HttpServletDispatcher.java:55)
	at org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher.service(HttpServletDispatcher.java:51)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:614)
	at io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:74)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:129)
	at org.keycloak.testsuite.UndertowRequestFilter.lambda$doFilter$0(UndertowRequestFilter.java:54)
	at org.keycloak.models.utils.KeycloakModelUtils.lambda$runJobInTransaction$1(KeycloakModelUtils.java:263)
	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransactionWithResult(KeycloakModelUtils.java:386)
	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:262)
	at org.keycloak.testsuite.UndertowRequestFilter.doFilter(UndertowRequestFilter.java:50)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:67)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)
	at io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68)
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at io.undertow.servlet.handlers.RedirectDirHandler.handleRequest(RedirectDirHandler.java:68)
	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:117)
	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64)
	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
	at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
	at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.SendErrorPageHandler.handleRequest(SendErrorPageHandler.java:52)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:276)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:135)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:132)
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:256)
	at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:101)
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:393)
	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:852)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
	at org.xnio.XnioWorker$WorkerThreadFactory$1$1.run(XnioWorker.java:1282)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot construct instance of `java.util.LinkedHashMap` (although at least one Creator exists): no String-argument constructor/factory method to deserialize from String value ('[{"key":"aaa","value":"bbb"},{"key":"bbb","value":"ccc"}]')
 at [Source: UNKNOWN; byte offset: #UNKNOWN] (through reference chain: org.keycloak.services.clientpolicy.condition.ClientAttributesCondition$Configuration["attributes"])
	at com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:63)
	at com.fasterxml.jackson.databind.DeserializationContext.reportInputMismatch(DeserializationContext.java:1754)
	at com.fasterxml.jackson.databind.DeserializationContext.handleMissingInstantiator(DeserializationContext.java:1379)
	at com.fasterxml.jackson.databind.deser.std.StdDeserializer._deserializeFromString(StdDeserializer.java:311)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:454)
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:32)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:310)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:177)
	at com.fasterxml.jackson.databind.ObjectMapper._convert(ObjectMapper.java:4619)
	... 74 more
```

### Version

nightly (from 2024-09-30)

### Regression

- [X] The issue is a regression

### Expected behavior

It is possible to login to such realm

### Actual behavior

It is not possible and users are stuck

### How to Reproduce?

See steps to reproduce

### Anything else?

It might be an admin console issue as there is a test to add such condition with the integration test ( `ClientPoliciesConditionTest.testClientAttributesCondition`) but not 100% sure. Will require some investigation.

--- a/server-spi/src/main/java/org/keycloak/models/IdentityProviderMapperModel.java
+++ b/server-spi/src/main/java/org/keycloak/models/IdentityProviderMapperModel.java
@@ -17,17 +17,11 @@
 
 package org.keycloak.models;
 
-import com.fasterxml.jackson.core.type.TypeReference;
-import org.keycloak.util.JsonSerialization;
+import org.keycloak.models.utils.MapperTypeSerializer;
 
-import java.io.IOException;
 import java.io.Serializable;
-import java.util.Collections;
 import java.util.List;
 import java.util.Map;
-import java.util.stream.Collectors;
-
-import static java.util.Collections.emptyMap;
 
 /**
  * Specifies a mapping from broker login to user data.
@@ -37,9 +31,6 @@ import static java.util.Collections.emptyMap;
  */
 public class IdentityProviderMapperModel implements Serializable {
     public static final String SYNC_MODE = "syncMode";
-
-    private static final TypeReference<List<StringPair>> MAP_TYPE_REPRESENTATION = new TypeReference<List<StringPair>>() {
-    };
 
     protected String id;
     protected String name;
@@ -98,20 +89,7 @@ public class IdentityProviderMapperModel implements Serializable {
 
     public Map<String, List<String>> getConfigMap(String configKey) {
         String configMap = config.get(configKey);
-        if (configMap == null) {
-            return emptyMap();
-        }
-
-        try {
-            List<StringPair> map = JsonSerialization.readValue(configMap, MAP_TYPE_REPRESENTATION);
-            return map.stream().collect(
-                    Collectors.collectingAndThen(
-                            Collectors.groupingBy(StringPair::getKey,
-                                    Collectors.mapping(StringPair::getValue, Collectors.toUnmodifiableList())),
-                            Collections::unmodifiableMap));
-        } catch (IOException e) {
-            throw new RuntimeException("Could not deserialize json: " + configMap, e);
-        }
+        return MapperTypeSerializer.deserialize(configMap);
     }
 
     @Override
@@ -129,26 +107,5 @@ public class IdentityProviderMapperModel implements Serializable {
     @Override
     public int hashCode() {
         return id.hashCode();
-    }
-
-    static class StringPair {
-        private String key;
-        private String value;
-
-        public String getKey() {
-            return key;
-        }
-
-        public void setKey(String key) {
-            this.key = key;
-        }
-
-        public String getValue() {
-            return value;
-        }
-
-        public void setValue(String value) {
-            this.value = value;
-        }
     }
 }

--- a/server-spi/src/main/java/org/keycloak/models/utils/MapperTypeSerializer.java
+++ b/server-spi/src/main/java/org/keycloak/models/utils/MapperTypeSerializer.java
@@ -1,0 +1,106 @@
+/*
+ * Copyright 2024 Red Hat, Inc. and/or its affiliates
+ *  and other contributors as indicated by the @author tags.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+package org.keycloak.models.utils;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import org.keycloak.util.JsonSerialization;
+
+import static java.util.Collections.emptyMap;
+
+/**
+ * Serializer and deserializer for {@link org.keycloak.provider.ProviderConfigProperty#MAP_TYPE}
+ *
+ * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
+ */
+public class MapperTypeSerializer {
+
+    private static final TypeReference<List<StringPair>> MAP_TYPE_REPRESENTATION = new TypeReference<>() {
+    };
+
+    public static Map<String, List<String>> deserialize(String configString) {
+        if (configString == null) {
+            return emptyMap();
+        }
+
+        try {
+            List<StringPair> map = JsonSerialization.readValue(configString, MAP_TYPE_REPRESENTATION);
+            return map.stream().collect(
+                    Collectors.collectingAndThen(
+                            Collectors.groupingBy(StringPair::getKey,
+                                    Collectors.mapping(StringPair::getValue, Collectors.toUnmodifiableList())),
+                            Collections::unmodifiableMap));
+        } catch (IOException e) {
+            throw new RuntimeException("Could not deserialize json: " + configString, e);
+        }
+    }
+
+    public static String serialize(Map<String, List<String>> config) {
+        List<StringPair> pairs = config.entrySet()
+                .stream()
+                .flatMap(entry -> {
+                    String key = entry.getKey();
+                    List<String> values = entry.getValue();
+                    return values
+                            .stream()
+                            .map(value -> new StringPair(key, value));
+                })
+                .toList();
+        try {
+            return JsonSerialization.writeValueAsString(pairs);
+        } catch (IOException e) {
+            throw new RuntimeException("Could not serialize json: " + config, e);
+        }
+    }
+
+    static class StringPair {
+        private String key;
+        private String value;
+
+        public StringPair() {
+        }
+
+        private StringPair(String key, String value) {
+            this.key = key;
+            this.value = value;
+        }
+
+        public String getKey() {
+            return key;
+        }
+
+        public void setKey(String key) {
+            this.key = key;
+        }
+
+        public String getValue() {
+            return value;
+        }
+
+        public void setValue(String value) {
+            this.value = value;
+        }
+    }
+}

--- a/server-spi/src/test/java/org/keycloak/utils/MapperTypeSerializerTest.java
+++ b/server-spi/src/test/java/org/keycloak/utils/MapperTypeSerializerTest.java
@@ -1,0 +1,75 @@
+/*
+ * Copyright 2024 Red Hat, Inc. and/or its affiliates
+ *  and other contributors as indicated by the @author tags.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+package org.keycloak.utils;
+
+import java.util.List;
+import java.util.Map;
+
+import org.junit.Assert;
+import org.junit.Test;
+import org.keycloak.common.util.MultivaluedHashMap;
+import org.keycloak.common.util.MultivaluedMap;
+import org.keycloak.models.utils.MapperTypeSerializer;
+
+/**
+ * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
+ */
+public class MapperTypeSerializerTest {
+
+    @Test
+    public void testBasicSerializeAndDeserialize() {
+        // Serialize
+        MultivaluedMap<String, String> simpleMap = new MultivaluedHashMap<>() {
+            {
+                putSingle("attr1", "Apple");
+                putSingle("attr2", "Orange");
+            }
+        };
+        String s = MapperTypeSerializer.serialize(simpleMap);
+
+        // Check after deserialize, it is equal to serialized
+        Map<String, List<String>> deserialized = MapperTypeSerializer.deserialize(s);
+        Assert.assertEquals(simpleMap, deserialized);
+
+        // Deserialize from String
+        deserialized = MapperTypeSerializer.deserialize("[{\"key\":\"attr2\",\"value\":\"Orange\"},{\"key\":\"attr1\",\"value\":\"Apple\"}]");
+        Assert.assertEquals(simpleMap, deserialized);
+    }
+
+    @Test
+    public void testMultivaluedSerializeAndDeserialize() {
+        // Deserialize with some multivalued value
+        Map<String, List<String>> deserialized = MapperTypeSerializer.deserialize("[{\"key\":\"attr2\",\"value\":\"Orange\"},{\"key\":\"attr1\",\"value\":\"Apple\"},{\"key\":\"attr2\",\"value\":\"Peach\"}]");
+        Assert.assertEquals(deserialized.get("attr1").size(), 1);
+        Assert.assertEquals(deserialized.get("attr1").get(0), "Apple");
+        Assert.assertEquals(deserialized.get("attr2").size(), 2);
+        Assert.assertTrue(deserialized.get("attr2").contains("Orange"));
+        Assert.assertTrue(deserialized.get("attr2").contains("Peach"));
+        Assert.assertFalse(deserialized.get("attr2").contains("Apple"));
+
+        // Serialize and deserialize again from String and check it is same value
+        String s = MapperTypeSerializer.serialize(deserialized);
+        Assert.assertEquals(MapperTypeSerializer.deserialize(s), deserialized);
+    }
+
+
+
+
+}

--- a/services/src/main/java/org/keycloak/services/clientpolicy/condition/ClientAttributesCondition.java
+++ b/services/src/main/java/org/keycloak/services/clientpolicy/condition/ClientAttributesCondition.java
@@ -20,12 +20,14 @@ package org.keycloak.services.clientpolicy.condition;
 import org.jboss.logging.Logger;
 import org.keycloak.models.ClientModel;
 import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.utils.MapperTypeSerializer;
 import org.keycloak.representations.idm.ClientPolicyConditionConfigurationRepresentation;
 import org.keycloak.services.clientpolicy.ClientPolicyContext;
 import org.keycloak.services.clientpolicy.ClientPolicyException;
 import org.keycloak.services.clientpolicy.ClientPolicyVote;
 import org.keycloak.services.clientpolicy.context.PreAuthorizationRequestContext;
 
+import java.util.List;
 import java.util.Map;
 
 /**
@@ -46,13 +48,13 @@ public class ClientAttributesCondition extends AbstractClientPolicyConditionProv
 
     public static class Configuration extends ClientPolicyConditionConfigurationRepresentation {
 
-        protected Map<String, String> attributes;
+        private String attributes;
 
-        public Map<String, String> getAttributes() {
+        public String getAttributes() {
             return attributes;
         }
 
-        public void setAttributes(Map<String, String> attributes) {
+        public void setAttributes(String attributes) {
             this.attributes = attributes;
         }
     }
@@ -100,7 +102,7 @@ public class ClientAttributesCondition extends AbstractClientPolicyConditionProv
     private boolean isAttributesMatched(ClientModel client) {
         if (client == null) return false;
 
-        Map<String, String> attributesForMatching = getAttributesForMatching();
+        Map<String, List<String>> attributesForMatching = getAttributesForMatching();
         if (attributesForMatching == null) return false;
 
         Map<String, String> clientAttributes = client.getAttributes();
@@ -111,12 +113,28 @@ public class ClientAttributesCondition extends AbstractClientPolicyConditionProv
         }
 
         return attributesForMatching.entrySet().stream()
-                .allMatch(entry -> clientAttributes.containsKey(entry.getKey()) && clientAttributes.get(entry.getKey()).equals(entry.getValue()));
+                .allMatch(entry -> {
+                    String key = entry.getKey();
+                    if (key == null) {
+                        logger.warnf("Empty key in configuration of client-attributes condition");
+                        return false;
+                    }
+                    if (entry.getValue() == null || entry.getValue().isEmpty()) {
+                        logger.warnf("Empty value in the configuration of client-attributes condition for the attribute %s. This cannot match any client", key);
+                        return false;
+                    }
+                    if (entry.getValue().size() > 1) {
+                        logger.warnf("More values in the configuration of client-attributes condition for the attribute %s. This cannot match any client", key);
+                        return false;
+                    }
+                    String value = entry.getValue().get(0);
+                    return clientAttributes.containsKey(key) && clientAttributes.get(key).equals(value);
+                });
     }
 
-    private Map<String, String> getAttributesForMatching() {
+    private Map<String, List<String>> getAttributesForMatching() {
         if (configuration.getAttributes() == null) return null;
-        return configuration.getAttributes();
+        return MapperTypeSerializer.deserialize(configuration.getAttributes());
     }
 
 }

--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/policies/ClientPoliciesConditionTest.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/policies/ClientPoliciesConditionTest.java
@@ -46,6 +46,7 @@ import org.keycloak.authentication.authenticators.client.JWTClientSecretAuthenti
 import org.keycloak.authentication.authenticators.client.X509ClientAuthenticator;
 import org.keycloak.client.registration.ClientRegistrationException;
 import org.keycloak.common.Profile;
+import org.keycloak.common.util.MultivaluedHashMap;
 import org.keycloak.models.AdminRoles;
 import org.keycloak.models.Constants;
 import org.keycloak.models.OAuth2DeviceConfig;
@@ -399,10 +400,10 @@ public class ClientPoliciesConditionTest extends AbstractClientPoliciesTest {
         json = (new ClientPoliciesBuilder()).addPolicy(
                 (new ClientPolicyBuilder()).createPolicy(POLICY_NAME, "Het Eerste Beleid", Boolean.TRUE)
                         .addCondition(ClientAttributesConditionFactory.PROVIDER_ID,
-                                createClientAttributesConditionConfig(new HashMap<String, String>() {
+                                createClientAttributesConditionConfig(new MultivaluedHashMap<String, String>() {
                                     {
-                                        put("attr1", "Apple");
-                                        put("attr2", "Orange");
+                                        putSingle("attr1", "Apple");
+                                        putSingle("attr2", "Orange");
                                     }
                                 }))
                         .addProfile(PROFILE_NAME)

--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/util/ClientPoliciesUtil.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/util/ClientPoliciesUtil.java
@@ -22,6 +22,7 @@ import com.fasterxml.jackson.databind.JsonNode;
 import com.fasterxml.jackson.databind.ObjectMapper;
 
 import org.keycloak.common.util.Base64Url;
+import org.keycloak.common.util.MultivaluedMap;
 import org.keycloak.crypto.AsymmetricSignatureSignerContext;
 import org.keycloak.crypto.ECDSASignatureSignerContext;
 import org.keycloak.crypto.KeyType;
@@ -33,6 +34,7 @@ import org.keycloak.jose.jwk.ECPublicJWK;
 import org.keycloak.jose.jwk.JWK;
 import org.keycloak.jose.jwk.RSAPublicJWK;
 import org.keycloak.jose.jws.JWSHeader;
+import org.keycloak.models.utils.MapperTypeSerializer;
 import org.keycloak.protocol.oidc.grants.ciba.clientpolicy.executor.SecureCibaAuthenticationRequestSigningAlgorithmExecutor;
 import org.keycloak.representations.dpop.DPoP;
 import org.keycloak.representations.idm.ClientPoliciesRepresentation;
@@ -406,9 +408,10 @@ public final class ClientPoliciesUtil {
         return config;
     }
 
-    public static ClientAttributesCondition.Configuration createClientAttributesConditionConfig(Map<String, String> attributes) {
+    public static ClientAttributesCondition.Configuration createClientAttributesConditionConfig(MultivaluedMap<String, String> attributes) {
         ClientAttributesCondition.Configuration config = new ClientAttributesCondition.Configuration();
-        config.setAttributes(attributes);
+        String attrsAsString = MapperTypeSerializer.serialize(attributes);
+        config.setAttributes(attrsAsString);
         return config;
     }
 

